### ISSUE TITLE
Proper scaling of mixed integer design variables

### ISSUE THREAD

**Issue by SebastianEEE** [CMA-ES/pycma#297]
Hello,

first of all: Thanks for the awesome package and the continued support and development! 

I am using pycma to solve a mixed integer optimization problem with design space bounds. Below is my exemplary implementation. I have read, that when using the option ´integer_variables´, I need to apply np.round() before actually evaluating my objective and constraint functions, which render the design variables to be integer-valued -2.0, -1.0, 0.0, 1.0., 2.0 etc. 

`import cma
import numpy as np

def obj_eval(des_var: np.array) -> float:
    # Round design variables where integer_design_variables_bool is True
    # When specifying integer_indices in options of CMA.fmin2,
    # the solver still proposes continuous design variables (e.g 3.67),
    # but it adapts its internal functioning compared to pure continuous working.
    # Thus we still need to round them to integers before evaluating our response function.
    for i, is_integer in enumerate(integer_design_variables_bool):
        if is_integer:
            des_var[i] = np.round(des_var[i])
            
    # des_var is proposed by optimizer within upper and lower bounds.
    # des_var holds unscaled values.
    output: float = some_function_using_unscaled_desvar(des_var)
    return output

def constraint_eval(des_var: np.array) -> float:
    # Round design variables where integer_design_variables_bool is True
    # When specifying integer_indices in options of CMA.fmin2,
    # the solver still proposes continuous design variables (e.g 3.67),
    # but it adapts its internal functioning compared to pure continuous working.
    # Thus we still need to round them to integers before evaluating our response function.
    for i, is_integer in enumerate(integer_design_variables_bool):
        if is_integer:
            des_var[i] = np.round(des_var[i])
            
    # des_var is proposed by optimizer within upper and lower bounds.
    # des_var holds unscaled values.
    output: float = some_function_using_unscaled_desvar(des_var)
    return output

integer_design_variables_bool = [False, False, True, True]
ub =                            [10,    100,   10,   100]
lb =                            [0,     0,     0,    0,]
x0 =                            [5,     50,    5,    50]

cfun = cma.ConstrainedFitnessAL(obj_eval, constraint_eval)

xopt, es_b = cma.fmin2(objective_function=cfun,
                                          x0=x0,
                                          sigma0=1,
                                          options={'bounds': [lb, ub],
                                                   'maxfevals': 100000,
                                                   'maxiter': 1000,
                                                   'tolfun': 1e-8,
                                                   'tolx': 1e-8,
                                                   'seed': 1,
                                                   'AdaptSigma': True,
                                                   'CMA_elitist': False,
                                                   'integer_variables': [i for i, is_integer in enumerate(integer_design_variables_bool) if is_integer]
                                                   },
                                          callback=cfun.update)
xopt = cfun.find_feasible(es_b)
for i, is_integer in enumerate(integer_design_variables_bool):
    if is_integer:
        xopt[i] = np.round(xopt[i])`

The code as above works fine for me, however I wanted to improve the numerical performance by scaling the design variables ranges to within [0, 1] for all four design variables (regardless of whether they are integer or continuous valued). 
Here is my suggestion:

`import cma
import numpy as np
from sklearn.preprocessing import MinMaxScaler

def obj_eval(des_var: np.array) -> float:
    for i, is_integer in enumerate(integer_design_variables_bool):
        if is_integer:
            des_var[i] = np.round(des_var[i])  # What to do about this ???
                                               # Possibly make use of 'decimals' option of np.round()
            
    # des_var is proposed by optimizer within upper and lower bounds.
    # des_var is now scaled within [0,1]
    # need to unscale before evaluation
    des_var = [scalers[i].inverse_transform([des_var[i]]) for i in range(len(des_var))]
    output: float = some_function_using_unscaled_desvar(des_var)
    return output

def constraint_eval(des_var: np.array) -> float:
    for i, is_integer in enumerate(integer_design_variables_bool):
        if is_integer:
            des_var[i] = np.round(des_var[i])   # What to do about this ???
                                                # Possibly make use of 'decimals' option of np.round()
            
    # des_var is proposed by optimizer within upper and lower bounds.
    # des_var is now scaled within [0,1]
    # need to unscale before evaluation
    des_var = [scalers[i].inverse_transform([des_var[i]]) for i in range(len(des_var))]
    output: float = some_function_using_unscaled_desvar(des_var)
    return output

integer_design_variables_bool = [False, False, True, True]
ub =                            [10,    100,   10,   100]
lb =                            [0,     0,     0,    0,]
x0 =                            [5,     50,    5,    50]

scalers =                     = [MinMaxScaler(feature_range=(0, 1).fit([[lb[0]], [ub[0]]])),
                                 MinMaxScaler(feature_range=(0, 1).fit([[lb[1]], [ub[1]]])),
                                 MinMaxScaler(feature_range=(0, 1).fit([[lb[2]], [ub[2]]])),
                                 MinMaxScaler(feature_range=(0, 1).fit([[lb[3]], [ub[3]]]))]

cfun = cma.ConstrainedFitnessAL(obj_eval, constraint_eval)

xopt, es_b = cma.fmin2(objective_function=cfun,
                                          x0=[scalers[i].transform([x0[i]]) for i in range(len(x0))],  # now x0 is within [0,1]
                                          sigma0=1,
                                          options={'bounds': [[0,0,0,0],
                                                              [1,1,1,1]],  # bounds are now within [0,1]
                                                   'maxfevals': 100000,
                                                   'maxiter': 1000,
                                                   'tolfun': 1e-8,
                                                   'tolx': 1e-8,
                                                   'seed': 1,
                                                   'AdaptSigma': True,
                                                   'CMA_elitist': False,
                                                   'integer_variables': [i for i, is_integer in enumerate(integer_design_variables_bool) if is_integer]
                                                   },
                                          callback=cfun.update)
xopt = cfun.find_feasible(es_b)
for i, is_integer in enumerate(integer_design_variables_bool):
    if is_integer:
        xopt[i] = np.round(xopt[i])  # What to do about this ???
                                     # Possibly make use of 'decimals' option of np.round()`

This approach should work fine for the continuous design variables: They are simply unscaled back to their original scale within the objective and constraint function. However, what to do about those design variables that are integer valued in their original scale? If I still apply np.round() they would all be rounded to 0 or 1 , when I actually need them to be rounded to their nearest 0.1 or 0.01 point instead (depending on their original unscaled range). I could just make use of the 'decimals' option of np.round(). But I am in general not sure, whether pycma's built-in 'integer_varriables' option still behaves as intended in that case? 

Please excuse the lengthy question and thank you very much for your help!

**Comment by nikohansen** [comment]
With the current implementation, it does not make sense to rescale integer variables (and in particular not into [0, 1]). This is not compatible with the integer handling. Instead, and equivalent with rescaling but valid, it is possible to set their initial standard deviations accordingly. This is in particular necessary when the remaining search space is scaled into [0, 1] (which may or may not make sense). This can be done with the `'CMA_stds'` option:

```python
import  cma
cma.CMAOptions('CMA_std')
```
```
{'CMA_stds': 'None  # multipliers for sigma0 in each coordinate (not represented in C), or use `cma.ScaleCoordinates` instead',
 'CMA_teststds': 'None  # factors for non-isotropic initial distr. of C, mainly for test purpose, see CMA_stds for production',
 'scaling_of_variables': 'None  # deprecated, rather use fitness_transformations.ScaleCoordinates instead (or CMA_stds). WAS: Scale for each variable in that effective_sigma0 = sigma0*scaling. Internally the variables are divided by scaling_of_variables and sigma is unchanged, default is `np.ones(N)`'}
```
Rescaling the bounds to [0, 1] and the initial point is _equivalent_ with setting the components of `'CMA_stds'` to the original boundary ranges, that is, the inverse of the rescale factor. The initial `sigma` then needs to be small (like <1/2), like for searching in the domain [0, 1].

**Comment by SebastianEEE** [comment]
Thank you for your quick answers. So you are suggesting to retain the original (unscsaled) search space and instead adapt 'CMA_stds'   as below? 
```
import cma
import numpy as np


def obj_eval(des_var: np.array) -> float:
    # Round design variables where integer_design_variables_bool is True
    # When specifying integer_indices in options of CMA.fmin2,
    # the solver still proposes continuous design variables (e.g 3.67),
    # but it adapts its internal functioning compared to pure continuous working.
    # Thus we still need to round them to integers before evaluating our response function.
    for i, is_integer in enumerate(integer_design_variables_bool):
        if is_integer:
            des_var[i] = np.round(des_var[i])
            
    # des_var is proposed by optimizer within upper and lower bounds.
    # des_var holds unscaled values.
    output: float = some_function_using_unscaled_desvar(des_var)
    return output

def constraint_eval(des_var: np.array) -> float:
    # Round design variables where integer_design_variables_bool is True
    # When specifying integer_indices in options of CMA.fmin2,
    # the solver still proposes continuous design variables (e.g 3.67),
    # but it adapts its internal functioning compared to pure continuous working.
    # Thus we still need to round them to integers before evaluating our response function.
    for i, is_integer in enumerate(integer_design_variables_bool):
        if is_integer:
            des_var[i] = np.round(des_var[i])
            
    # des_var is proposed by optimizer within upper and lower bounds.
    # des_var holds unscaled values.
    output: float = some_function_using_unscaled_desvar(des_var)
    return output

integer_design_variables_bool = [False, False, True, True]
ub =                            [10,    100,   10,   100]
lb =                            [0,     0,     0,    0,]
x0 =                            [5,     50,    5,    50]

cfun = cma.ConstrainedFitnessAL(obj_eval, constraint_eval)

xopt, es_b = cma.fmin2(objective_function=cfun,
                                          x0=x0,
                                          sigma0=1/3,
                                          options={'bounds': [lb, ub],
                                                   'CMA_stds': [u-l for l,u in zip(lb,ub)]
                                                   'maxfevals': 100000,
                                                   'maxiter': 1000,
                                                   'tolfun': 1e-8,
                                                   'tolx': 1e-8,
                                                   'seed': 1,
                                                   'AdaptSigma': True,
                                                   'CMA_elitist': False,
                                                   'integer_variables': [i for i, is_integer in enumerate(integer_design_variables_bool) if is_integer]
                                                   },
                                          callback=cfun.update)
xopt = cfun.find_feasible(es_b)
for i, is_integer in enumerate(integer_design_variables_bool):
    if is_integer:
        xopt[i] = np.round(xopt[i])
```

**Comment by nikohansen** [comment]
> So you are suggesting to retain the original (unscaled) search space and instead adapt 'CMA_stds' as below?

yes.
